raiden/transfer/log.py

```diff
@@ -0,0 +1,238 @@
+# -*- coding: utf-8 -*-
+import pickle
+import sqlite3
+from abc import ABCMeta, abstractmethod
+
+from raiden.utils import create_file_iff_not_existing
+
+
+# TODO:
+# - snapshots should be used to reduce the log file size
+
+class TransactionLogSerializer(object):
+    """ TransactionLogSerializer
+
+        An abstract class defining the serialization interface for the
+        Transaction log. Allows for pluggable serializer backends.
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def serialize(self, transaction):
+        pass
+
+    @abstractmethod
+    def deserialize(self, data):
+        pass
+
+
+class PickleTransactionSerializer(TransactionLogSerializer):
+    """ PickleTransactionSerializer
+
+        A simple transaction serializer using pickle
+    """
+    def serialize(self, transaction):
+        # Some of our StateChange classes have __slots__ without having a __getstate__
+        # As seen in the SO question below:
+        # http://stackoverflow.com/questions/2204155/why-am-i-getting-an-error-about-my-class-defining-slots-when-trying-to-pickl#2204702
+        # We can either add a __getstate__ to all of them or use the `-1` protocol and be
+        # incompatible with ancient python version. Here I opt for the latter.
+        return pickle.dumps(transaction, -1)
+
+    def deserialize(self, data):
+        return pickle.loads(data)
+
+
+class TransactionLogStorageBackend(object):
+    """ TransactionLogStorageBackend
+
+        An abstract class defining the storage backend for the transaction log.
+        Allows for pluggable storage backends.
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def write_transaction(self, identifier, data):
+        pass
+
+    @abstractmethod
+    def write_state_snapshot(self, identifier, data):
+        pass
+
+    @abstractmethod
+    def read(self):
+        pass
+
+    @abstractmethod
+    def last_identifier(self):
+        pass
+
+
+class TransactionLogSQLiteBackend(TransactionLogStorageBackend):
+
+    def __init__(self, database_path):
+        self.conn = sqlite3.connect(database_path)
+        self.conn.text_factory = str
+        cursor = self.conn.cursor()
+        cursor.execute(
+            'CREATE TABLE IF NOT EXISTS transactions ('
+            '    id integer primary key autoincrement, data binary'
+            ')'
+        )
+        cursor.execute(
+            'CREATE TABLE IF NOT EXISTS state_snapshot (id integer primary key, data binary)'
```

Comment:
Yes that is true. The snapshot is not implemented as part of this PR though, this is just a skeleton for the snapshot, but I can make this modification in the DB scheme with this PR.